integrationtests/agent/adoption_test.go

```diff
@@ -0,0 +1,387 @@
+					return fmt.Errorf("key %s not found or value %s does not match %s in %+v", k, v2, v, super)
+				}
+			}
+		}
+		return nil
+	}
+
+	// configMapAdoptedAndMerged checks that the ConfigMap is adopted. It may
+	// need to be extended to check for more labels and annotations.
+	isConfigMapAdopted := func(cm *corev1.ConfigMap) error {
+		err := mapPartialMatch(cm.Labels, map[string]string{
+			"app.kubernetes.io/managed-by": "Helm",
+		})
+		if err != nil {
+			return err
+		}
+		err = mapPartialMatch(cm.Annotations, map[string]string{
+			"meta.helm.sh/release-name":      "",
+			"meta.helm.sh/release-namespace": "",
+		})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	changeConfigMap := func(name string, change func(*corev1.ConfigMap)) {
+		cm := &corev1.ConfigMap{}
+		Eventually(func() error {
+			return k8sClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, cm)
+		}).Should(Succeed())
+		change(cm)
+		Eventually(func() error {
+			return k8sClient.Update(ctx, cm)
+		}).Should(Succeed())
+	}
+
+	deleteConfigMap := func(name string) {
+		cm := &corev1.ConfigMap{}
+		Eventually(func() error {
+			return k8sClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, cm)
+		}).Should(Succeed())
+		Eventually(func() error {
+			return k8sClient.Delete(ctx, cm)
+		}).Should(Succeed())
+	}
+
+	bundleDeploymentResourceMissing := func(bd *v1alpha1.BundleDeployment) error {
+		const msgPrefix = "BundleDeployment resource:"
+		for _, condition := range bd.Status.Conditions {
+			if condition.Type != v1alpha1.BundleDeploymentConditionReady {
+				continue
+			}
+			if condition.Status != corev1.ConditionFalse {
+				return fmt.Errorf("%s Status is not False", msgPrefix)
+			}
+			if condition.Reason != "Error" {
+				return fmt.Errorf("%s Reason is not Error", msgPrefix)
+			}
+			if !strings.Contains(condition.Message, "missing") {
+				return fmt.Errorf("%s Message does not contain 'missing'", msgPrefix)
+			}
+		}
+		return nil
+	}
+
+	bundleDeploymentNotOwnedByUs := func(bd *v1alpha1.BundleDeployment) error {
+		for _, condition := range bd.Status.Conditions {
+			if condition.Type == v1alpha1.BundleDeploymentConditionReady &&
+				condition.Status == corev1.ConditionFalse &&
+				condition.Reason == "Error" &&
+				strings.Contains(condition.Message, "not owned by us") {
+				return nil
+			}
+		}
+		return fmt.Errorf("does not match expected condition")
+	}
+
+	bundleDeploymentReady := func(bd *v1alpha1.BundleDeployment) error {
+		for _, condition := range bd.Status.Conditions {
+			if condition.Type == v1alpha1.BundleDeploymentConditionReady &&
+				condition.Status == corev1.ConditionTrue {
+				return nil
+			}
+		}
+		return fmt.Errorf("does not match expected condition")
+	}
+
+	BeforeEach(func() {
+		namespace = createNamespace()
+		DeferCleanup(func() {
+			Expect(k8sClient.Delete(ctx, &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{Name: namespace}})).ToNot(HaveOccurred())
+		})
```

Comment:
I like the intention, but is this really necessary, considering that integration tests don't run on an actual cluster and that no cleanup will happen based on ownership references, etc ([source](https://book.kubebuilder.io/reference/envtest#testing-considerations))?

Reply:
I think it's even worse than that. From your referenced source:

> EnvTest does not support namespace deletion. Deleting a namespace will seem to succeed, but the namespace will just be put in a Terminating state, and never actually be reclaimed.

For which reason every tests is assigned its own namespace.

I just didn't think much about it when I copied it from one of the other tests next to this one, but now that I do, I feel like I would have to "fix" it everywhere.